mcs/class/System/System/MonoExeLocator.cs

```diff
@@ -0,0 +1,82 @@
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace System {
+
+	public static class MonoExeLocator {
+
+		public static string GacPath { get; }
+		public static string MonoPath { get; }
+		public static string McsPath { get; }
+		public static string VbncPath { get; }
+		
+		static MonoExeLocator () {
+
+			PropertyInfo gac = typeof (Environment).GetProperty ("GacPath", BindingFlags.Static | BindingFlags.NonPublic);
+			MethodInfo get_gac = gac.GetGetMethod (true);
+			string p = Path.GetDirectoryName ((string) get_gac.Invoke (null, null));
+			GacPath = p;
+
+			string monoPath = null;
+			string mcsPath = null;
+			string vbncPath = null;
+
+			if (Path.DirectorySeparatorChar == '\\') {
+				string processExe = Process.GetCurrentProcess ().MainModule.FileName;
+				if (processExe != null) {
+					string fileName = Path.GetFileName (processExe);
+					if (fileName.StartsWith ("mono") && fileName.EndsWith (".exe"))
+						monoPath = processExe;
+				}
+
+				if (!File.Exists (monoPath))
+					monoPath = Path.Combine (
+						Path.GetDirectoryName (
+							Path.GetDirectoryName (p)),
```

Comment:
nit: use `GacPath` here instead of `p`, or rename `p` to something more descriptive.